File: Treachery-iOS/Treachery-iOS/Models/FriendRequest.swift
//
//  FriendRequest.swift
//  Treachery-iOS
//
//  Created by Luke Solomon on 3/14/26.
//

import Foundation

struct FriendRequest: Codable, Identifiable {
    var id: String
    let fromUserId: String
    let fromDisplayName: String
    let toUserId: String
    let status: FriendRequestStatus
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case fromUserId = "from_user_id"
        case fromDisplayName = "from_display_name"
        case toUserId = "to_user_id"
        case status
        case createdAt = "created_at"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id) ?? ""
        fromUserId = try container.decode(String.self, forKey: .fromUserId)
        fromDisplayName = try container.decode(String.self, forKey: .fromDisplayName)
        toUserId = try container.decode(String.self, forKey: .toUserId)
        status = try container.decode(FriendRequestStatus.self, forKey: .status)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
    }

    init(
        id: String,
        fromUserId: String,
        fromDisplayName: String,
        toUserId: String,
        status: FriendRequestStatus,
        createdAt: Date
    ) {
        self.id = id
        self.fromUserId = fromUserId
        self.fromDisplayName = fromDisplayName
        self.toUserId = toUserId
        self.status = status
        self.createdAt = createdAt
    }
}

enum FriendRequestStatus: String, Codable {
    case pending
    case accepted
    case declined
}

File: TreacheryAndroid/app/src/main/java/com/solomon/treachery/model/FriendRequest.kt
package com.solomon.treachery.model

import com.google.firebase.Timestamp

enum class FriendRequestStatus(val value: String) {
    PENDING("pending"),
    ACCEPTED("accepted"),
    DECLINED("declined");

    companion object {
        fun fromValue(value: String): FriendRequestStatus? = entries.find { it.value == value }
    }
}

data class FriendRequest(
    val id: String = "",
    val fromUserId: String = "",
    val fromDisplayName: String = "",
    val toUserId: String = "",
    val status: FriendRequestStatus = FriendRequestStatus.PENDING,
    val createdAt: Timestamp = Timestamp.now()
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "from_user_id" to fromUserId,
        "from_display_name" to fromDisplayName,
        "to_user_id" to toUserId,
        "status" to status.value,
        "created_at" to createdAt
    )

    companion object {
        fun fromMap(id: String, data: Map<String, Any?>): FriendRequest = FriendRequest(
            id = id,
            fromUserId = data["from_user_id"] as? String ?: "",
            fromDisplayName = data["from_display_name"] as? String ?: "",
            toUserId = data["to_user_id"] as? String ?: "",
            status = FriendRequestStatus.fromValue(data["status"] as? String ?: "") ?: FriendRequestStatus.PENDING,
            createdAt = data["created_at"] as? Timestamp ?: Timestamp.now()
        )
    }
}
